field/TinyLang/Field/Evaluator.hs

```diff
@@ -51,43 +56,85 @@ data EvalError f
     -- TODO: proper error messages.
     deriving (Show)
 
-type MonadEvalError f m = MonadError (EvalError f) m
+-- | Evaluator Transformer Stack
+newtype EvaluatorT r s e (m :: Type -> Type) a =
+    EvaluatorT { runEvaluatorT :: ReaderT r (StateT s (ExceptT e m)) a }
+    deriving newtype ( Monad
+                     , Functor
+                     , Applicative
+                     , MonadError e
+                     , MonadState s
+                     , MonadReader r
+                     )
+
+-- type Evaluator r s e a = EvaluatorT r s e Identity a
+-- runEvaluator :: Evaluator  r s e a -> ReaderT r (StateT s (Except e)) a
+-- runEvaluator = runEvaluatorT
+
+execEvaluatorT :: (Monad m) => EvaluatorT r s e m a -> r -> s -> m (Either e s)
+execEvaluatorT ev r s = runExceptT $ flip execStateT s $ flip runReaderT r $ runEvaluatorT ev
+
+-- execEvaluator :: Evaluator r s e a -> r -> s -> Either e s
+-- execEvaluator ev r s = runIdentity $ execEvaluatorT ev r s
+
+-- | Type of evaluator specialised to our case.  We wrapped it in newtype to
+-- improve error messages.
+newtype EvalT f m a =
+    EvalT { unEvalT :: EvaluatorT (Env (SomeUniConst f)) [SomeUniConst f] (EvalError f) m a
+          }
+    deriving newtype ( Monad
+                     , Functor
+                     , Applicative
+                     , MonadError (EvalError f)
+                     , MonadState [SomeUniConst f]
+                     , MonadReader (Env (SomeUniConst f))
+                     )
+
+type Eval f a = EvalT f Identity a
+
+execEvalT :: (Monad m) => EvalT f m a -> Env (SomeUniConst f) -> m (Either (EvalError f) [SomeUniConst f])
+execEvalT ev r = execEvaluatorT (unEvalT ev) r mempty
+
+execEval :: Eval f a -> Env (SomeUniConst f) -> Either (EvalError f) [SomeUniConst f]
+execEval ev r = runIdentity $ execEvalT ev r
 
 instance TextField f => Show (TypeMismatch f) where
     show (TypeMismatch (UniVar uniExp var) uniConstAct) =
         [QQ.i|Variable #{var} has type #{uniExp}, but was instantiated with #{uniConstAct}|]
 
-lookupVarEval :: MonadEvalError g m => Var -> Env f -> m f
-lookupVarEval var = fromOption (throwError $ VariableNotInScopeEvalError var) . lookupVar var
+lookupVarM :: (Monad m) => Var -> EvalT f m (SomeUniConst f)
+lookupVarM var = do
+    env <- ask
+    fromOption (throwError $ VariableNotInScopeEvalError var) $ lookupVar var env
 
-invEval :: (MonadEvalError g m, Field f) => f -> m f
+invEval :: (Monad m, Field f) => f -> EvalT g m f
```

Comment:
At this point `EvalT` implements `MonadEvalError`.  Just for clarification:  should I update all the exported functions to use the `mtl`-syle `MonadEvalError`?  (Or probably more aptly named `MonadEval`?)

Reply:
What I mean is that functions that only throw and don't do anything with `Reader` or `State`, should run in the general `MonadEvalError`, because some of those functions are reused in the compiler. For example `invEval` doesn't do anything with `Reader` or `State`, hence it should stay generic and run in any `MonadEvalError`, which was defined as

```haskell
type MonadEvalError f m = MonadError (EvalError f) m
```

I.e. "any `m` is good enough as long as you can throw a `EvalError f` in it".

Functions that do stuff with `Reader` or `State` should stay the way they're implemented right now.